### ISSUE TITLE
Simple interface to specifying models in DiscreteDP

### ISSUE THREAD

**Issue by jstac** [QuantEcon/QuantEcon.py#228]
@oyamad  This is just a wish list idea that I'm noting down while I think of it.  (Perhaps other people have mentioned it too.)

It would be nice to have an interface to specifying discrete DP models for DiscreteDP by a simplified version of what @albop has built for dolo:

http://dolo.readthedocs.org/en/doc/modeling_language.html#example

The interface would allow you to easily specify

* the feasible state action pairs, 
* the transition probs
* the rewards for each state action pair

A numbafied routine would map that specification to an instance of DiscreteDP.

This code would be greatly simplified and more attractive to users: http://quant-econ.net/py/aiyagari.html

Same goes for the Julia side.

**Comment by oyamad** [comment]
@jstac Thanks for the suggestion. Let me think about it.

I suppose you are not suggesting we invent a "modeling language" as in dolo. Maybe we should start with a simple example.

**Comment by spencerlyon2** [comment]
We'll definitely need input from @albop on this, but to me it makes sense to _extend_ the modeling language of dolo to allow the input of `dtdsdc` (**d**iscrete **t**ime **d**iscrete **s**tate **d**discrete **c**ontrol) models that leverages this code as a compilation backend.

**Comment by spencerlyon2** [comment]
So I did some work and have a few changes to push. I drafted an email to Lars/Tom that explains what is going on and wanted to give you a chance to look at it before I hit send. Warning, it is super long, but here's the email.

- - - 

Dear Lars and Tom,

Thank you for your detailed response, that helped us very much. I apologize in advance, but this email will be quite long. We got some surprising results, so we want to ensure that we are all on the same page regarding how the computation is carried out.

Alberto and I have tried to implement the steps you outlined, but have some surprising results. 

## I. Experiment along transition path

To make our discussion more concrete, I will include the code snippet that defines the main loop over time in our simulation:

```
for t in 2:horizon+1
    i = t-1

    # compute the average output from all employed workers in our cohort of
    # agents who matched at the start of period 2. Note that this is the
    # same in both economies because the path of A and the measure of
    # unemployed workers is the same
    av_output[i] = trans_eff.Aτ[t]*dot(zgrid, dnu)

    # now, in the efficient wage and sticky wage cases, compute the average
    # wage cost. This will be simply equal to the inner product between the
    # wage at each productivity level and the distribution over
    # productivity levels for our cohort
    av_wcost_eff[i] = dot(Phi*trans_eff.cwτ[:, t], dnu)
    av_wcost_stk[i] = dot(Phi*trans_stk.cwτ[:, t], dnu)

    # compute the discount factor for each economy.
    # NOTE: We we are looking for the value in period 2 "units" (units is
    #       probably the wrong word here, but we want the base year for
    #       discounting to be year 2). So, we begin by setting the discount
    #       factor for period `t` to `β^(t-2) u'(C_{t})/u'(C_2)`. Notice
    #       that on the first iteration (time t=2, our starting period)
    #       this value is 1, which means that the flow profits in period 2 
    #       are not discounted. This is what we want in this experiment
    disc_eff[i] = β^(t-2)*(trans_eff.Cτ[t]/trans_eff.Cτ[2])^(-α)
    disc_stk[i] = β^(t-2)*(trans_stk.Cτ[t]/trans_stk.Cτ[2])^(-α)

    # NOTE: Because we are not including exogenous separation or worker
    #       death in the transition rule for the distribution of our cohort
    #       we will build those probabilities into the discount factor
    #       here. Thus, we will multiply the discount factor computed above
    #       by the probability that the a match made in period 2 survives to
    #       period t by multiplying by `(ϕ*(1-σ))^(t-2)`.  Note again that
    #       on the first iteration (t=2) this value is necessarily 1, which
    #       means that the matches made in period 2 produce with
    #       probability 1 in period 2. In other words, matches don't get
    #       formed and broken without any production happening. This is
    #       also what we want.
    disc_eff[i] *= (ϕ*(1-σ))^(t-2)
    disc_stk[i] *= (ϕ*(1-σ))^(t-2)

    # Compute the average present value of the match made in period 2. For
    # each economy, this is computed as the product of the effective
    # discount factor times the difference between average output and
    # average wage cost.
    pv_eff[i] = (av_output[i] - av_wcost_eff[i])*disc_eff[i]
    pv_stk[i] = (av_output[i] - av_wcost_stk[i])*disc_stk[i]

    # update the distribution of our cohort of agents who matched in period
    # 2. This will simply be the transition matrix times the current
    # distribution. This is the same for both models because dnu and Fzze
    # are both the same across models.
    # NOTE: we do this _without_ attrition and mortality. See above
    dnu = dnu'Fzze
end
```

A few comments about the code:

- Although probably clear from context, I'll point out that the `#` starts a comment.
- Variables with suffix `_eff` are from the efficient wage model with proportional home production benefits. Variables with the `_stk` suffix are from the corresponding sticky wage economy.
- Parameters `ϕ`, `σ`, `α`, and `β` are as in KMP and are the same in both economies.
- `zgrid` contains the grid over worker productivity levels.
- At the start of the loop, the variable `dnu` contains the steady state distribution of unemployed workers. That is, `dnu[i]` contains the mass of unemployed agents that have productivity `zgrid[i]` in the steady state. Note that because we have the actual distribution of workers and a transition law for updating the distribution, we did not do a monte carlo simulation of many workers but instead operated directly with the distribution.
- the `trans` object contains the data we collected when we performed the transition experiment of KMP. So 
    - `trans.Aτ` contains the time series of `A`
    - `trans.Qτ` contains the time series of one period ahead stochastic discount factors (see comment for details)
    - `trans.cwτ` contain the time series of coefficients for the wage at each period. `Phi` (the same in both models) is the spline basis matrix we use for interpolation, so to get the wage in period `t` along the transition path we compute `m.g.Phi * trans.cwτ[:, t]`
- Finally `Fzze` is a stochastic matrix that defines the transition for the distribution of workers moving from an employed state in period t to an employed state in period t+1. We use this at the bottom of the loop to update our `dnu` variable containing the distribution of our cohort of workers who matched in period 2.

### I.A Results

Now with the code in front of us, I show how we arrive at our main result.

In our simulation, the present discounted value in period 2 of creating a vacancy is the sum of either `pv_eff` in the efficient wage model or the sum of `pv_stk` in the sticky wage model. 

The parameter `κ` from KMP's paper is the cost of creating a vacancy that "drives the expected value of posting a vacancy to 0". To compute the implied cost of posting a vacancy, conditional on filling the vacancy we compute `κ/λf_2`, where `λf_2` is the probability that a vacancy is filled in period 2. 

Thus, to finish the experiment we computed the following:

```
# efficient wage model 
have_eff = sum(pv_eff)
want_eff = κ/trans_eff.λfτ[2]

# sticky wage model
have_stk = sum(pv_stk)
want_stk = κ/trans_stk.λfτ[2]
```

The numerical values are:

```
have_eff = 10.32703
want_eff = 9.947265

have_stk = 10.05883
want_stk = 9.984873
```

Notice that in both the efficient wage and sticky wage version of the model what we computed in our experiment and what should hold are not exactly equal. We get closer in the sticky wage model, but it is still not exact. 

## II. Experiment in steady state

We weren't sure if this was an issue of numerical accuracy in the items we computed in `trans_eff` and `trans_stk`, so we repeated the same experiment staying entirely in the steady state. The main loop for this case is:

```
for t in 2:horizon+1
    i = t-1

    # Keep A at it's SS value of 1.0
    av_output[i] = dot(zgrid, dnu)

    # use steady state wage
    av_wcost[i] = dot(bm.w, dnu)

    # C_{t+1} = C_t, so discount factor is just β^t. We also add back in
    # mortaility and job separation probabilities here.
    disc[i] = (β*ϕ*(1-σ))^(t-2)

    # unchanged
    pv[i] = (av_output[i] - av_wcost[i])*disc[i]
    dnu= dnu'Fzze
end
```

Comments on code:

- The steady state for both models is the same, so we only do computations for one model. 
- `bm.w` is the steady state wage
- 

### II.A Results

When I compute the corresponding `have` and `want` variables for this experiment I get

```
have = 10.443449
want = 10.056218
```

This result made us question whether we are doing one of the steps of this experiment incorrectly.

Do you see anything in our procedure that looks incorrect to you?

Thank you for helping with this and making it through this long email.

**Comment by spencerlyon2** [comment]
Please disregard that last comment (that has been deleted). I accidentally posted to the wrong repository. Sorry for the noise

**Comment by albop** [comment]
A quick feedback before I come back from holidays:
 
The model presented in the lecture corresponds to a dtcscc model for the partial eq case  so it can be defined using the existing dolo syntax, modulo some terminological adjustments. Then it would be relatively easy to write a translator/wrapper to translate to the few matrices that discretedp operates on. That would essentially be be another solution method, a good dolo application and a way to strengthen the links between the two libraries. In a similar vein, there is ongoing work to describe and solve the very same heterogeneous agents model in dolo. We expect to finish it by the end of next week.

There is a different problem which consists in defining a syntax for matrices and operations on them. There is some limited support for matrices in dolo/dtmscc models to define Markov matrices,  discretize shocks  and combine them together, but I have never found it completely convincing and have always wanted to find something more systematic and robust. Another way to construct Markov matrices (not in this example) would be to allow for conditional expressions which don't exist yet.

Basically depending on how transparent you want the construction of the model matrices to be in the lectures, depending on whether the model is discrete or discetize I would suggest one way or the other.

**Comment by oyamad** [comment]
[Human generated text] I had a discussion with Claude Code Fable 5 about this issue across libraries, QE.py, QE.jl, and [ContinuousDPs.jl](https://github.com/QuantEcon/ContinuousDPs.jl). Below is a copy-paste from the summary of that discussion generated by  Claude Code Fable 5.

---
Coming back to this after nearly a decade: the technical obstacle that stalled it (passing user functions into numba-jitted construction loops) is long gone, and this is a proposal for a concrete API. The scoping from the original thread stands: no modeling language on our side (a dolo→`DiscreteDP` translator belongs in dolo, as @albop suggested). The durable core of the request is narrower than a language: a modeling language represents the model symbolically, which buys symbolic services `DiscreteDP` never uses — the solvers only ever *evaluate* `R` and `Q` entries. For construction, evaluation access is the entire requirement, so passing plain functions (with parameters captured in closures) suffices; what QuantEcon.py needs is a programmatic interface — callables in, model out.

## Precedent, and a working prototype

Two things make this more than a design sketch.

First, [POMDPs.jl](https://github.com/JuliaPOMDP/POMDPs.jl) is a decade-tested instance of exactly this design: [QuickPOMDPs](https://juliapomdp.github.io/POMDPs.jl/latest/def_pomdp/) defines models by passing functions as keyword arguments, `TabularPOMDP` is the arrays form (a `DiscreteDP` sibling), and their explicit-vs-generative split marks the same boundary as the materialize-vs-implicit distinction discussed below. Notably, no modeling language emerged in that ecosystem — functions proved sufficient.

Second, QuantEcon/QuantEcon.jl#398 proposes a `DiscreteDP` ↔ POMDPs.jl bridge whose solver direction — enumerate states and feasible actions, query `transition`/`reward`, materialize `R`/`Q`, run our solvers — *is* the "construct a `DiscreteDP` from function definitions" builder, with the contract already designed by POMDPs.jl (`actions(m, s)` as the feasibility function, distribution objects as the transition contract, `stateindex` as the integer mapping). That bridge will serve as the working prototype for this proposal, and its lessons should flow back here before the Python API below is frozen.

## Proposed interface

A classmethod constructor, with a jitted two-pass builder behind it (count feasible pairs, allocate exactly, fill):

```python
ddp = DiscreteDP.from_functions(
    num_states,       # int
    num_actions,      # int: upper bound on the action index
    reward,           # callable: (s, a) -> float
    transition,       # callable: (s, a, out_states, out_probs) -> k, filling the
                      #   first k entries with the support and probabilities of Q(s, a)
    max_next_states,  # int: upper bound on k, sizing the fill buffers
    beta,
    feasible=None,    # callable: (s, a) -> bool; None means all pairs feasible
)
```

Design points:

- **Integer indices at the core.** The callables take state/action *indices*; model values (asset grids, shock values, prices) enter via closure, which keeps everything jittable through the standard closure-factory pattern and translates directly to the state-action-pairs formulation. Value-based convenience is a thin wrapper the user writes in two lines, not API surface.
- **Sparse-friendly transitions by construction.** The buffer-filling `transition` contract is the one deliberate departure from the POMDPs.jl contract (which returns distribution *objects*): variable-length returns and object construction are numba-hostile, so the buffer form is the right translation of the same semantics. `Q` materializes directly as CSR from `(s_indices, a_indices, data)` triplets — no dense intermediate.
- **Output is an ordinary `DiscreteDP`.** All existing solvers, input validation, and performance apply unchanged; this is purely a front-end.
- **Graceful degradation.** If the callables are jitted, the builder specializes and runs at nopython speed; if not, a plain-Python loop produces the same arrays (construction happens once, so even the slow path is usable for medium problems).

## What the motivating example becomes

The Aiyagari household setup cited when this issue was opened — the `DiscreteDP`-based lecture version, now archived (`aiyagari_household.py`), with its `populate_R`/`populate_Q` loops and hand-rolled index arithmetic — reduces to the model equations plus one call:

```python
@njit
def reward(s, a):
    a_i, z_i = divmod(s, z_size)
    return np.log(w * z_vals[z_i] + (1 + r) * a_vals[a_i] - a_vals[a])

@njit
def feasible(s, a):
    a_i, z_i = divmod(s, z_size)
    return w * z_vals[z_i] + (1 + r) * a_vals[a_i] - a_vals[a] > 0

@njit
def transition(s, a, out_states, out_probs):
    z_i = s % z_size
    for j in range(z_size):
        out_states[j] = a * z_size + j
        out_probs[j] = Pi[z_i, j]
    return z_size

ddp = DiscreteDP.from_functions(a_size * z_size, a_size, reward, transition,
                                max_next_states=z_size, beta=beta, feasible=feasible)
```

The subsequent history of that lecture is itself evidence for this proposal: the current Aiyagari lecture has moved to a hand-rolled JAX solver, and its successors still hand-write `populate_R`-style array builders — the construction boilerplate outlived `DiscreteDP` in the lectures, and lowering it is exactly what this interface does. The rewrite of the archived version is the acceptance test: if it is not dramatically shorter and clearer, the API is wrong.

## Performance scope, and the JAX migration

`from_functions` is deliberately performance-neutral: it changes how a `DiscreteDP` is built, not what the solvers do. That keeps the claims honest with respect to the lectures' migration to JAX. What the JAX Aiyagari wins on is only partly hardware; the larger factor is *structure*: the state factorizes as (assets, shock), the action determines the next assets deterministically, and the shock follows an exogenous Markov chain, so the expected value is one small matrix product (`V @ Pi.T`) plus a gather — `Q` is never materialized. A generic `DiscreteDP`, with or without this interface, cannot see that structure: it stores the same information as ~10^8 CSR entries where `Pi` has 100.

This points to a natural second stage, which this proposal deliberately does not include but is designed not to preclude — a *factorized* constructor that declares the structure as a type instead of flattening it away:

```python
model = DiscreteDP.from_exogenous_factorization(   # name TBD
    endog_size, exog_chain, num_actions,
    reward,        # (k, z, a) -> float, or an (n_k, n_z, n_a) array
    endog_next,    # (k, z, a) -> k'; default: the action is k'
    beta, feasible=None,
)
```

Its solvers use the two-stage kernels (one GEMM for the expected value, one vectorized gather/max for the Bellman update), its memory footprint drops from CSR gigabytes to `Pi` plus the reward array, and the kernels are backend-neutral: numba on CPU, or a JAX backend that is, in effect, the current lecture code behind a validated interface with policy iteration and error bounds. The workflow the lectures currently lack falls out: define the model once, validate small instances against the generic `DiscreteDP` solvers, scale the same model object on GPU. (For this model class the factorization also dissolves the memory wall of #185 without any implicit machinery.)

## Relation to #185 (item 1)

The callable contract of `from_functions` is deliberately identical to what an *implicit* (array-free) `DiscreteDP` needs: the solvers only ever require `r(s, a) + β Σ q(s'|s, a) v(s')` on the fly, plus a materialized `Q_σ` (one action per state — tiny) for policy evaluation. So `from_functions` is stage one of a two-backend design: **materialize** (this proposal — all existing machinery, for problems that fit in memory) and **implicit** (the #185 resolution — O(n) memory for problems that don't), with the user writing the model once and the backend chosen explicitly or by estimated size. This is the same boundary POMDPs.jl draws between its explicit `transition` and its generative `gen`. Designing the contract once closes both issues coherently.

## Open questions

1. Naming: `DiscreteDP.from_functions` vs a module-level factory.
2. Should infeasibility alternatively be signaled by `reward` returning `-inf` (the existing product-formulation convention), making `feasible` optional sugar rather than a separate callable?
3. Should `max_next_states` be inferred by a probing pass instead of required?
4. Sequencing: settle the design (or a prototype) of QuantEcon/QuantEcon.jl#398 first and port its lessons here, or develop the two in parallel?
5. Should the factorized constructor be co-designed now (at least its name and its relation to `from_functions`), or staged strictly after this proposal?

🤖 Generated with [Claude Code](https://claude.com/claude-code) (Claude Fable 5)

**Comment by jstac** [comment]
Thanks @oyamad 

The factorized constructor is nice --- it addressed the issue neatly by using available structure --- but seems rather specialized to the application in question.

If the package moves to JAX, which seems like a clear improvement, then one more general option is to let the user specify the high-dimensional linear operations involved in DiscreteDP as linear operators.  A relevant discussion is here:

https://jax.quantecon.org/markov_asset.html#a-memory-efficient-jax-version

This avoids instatiating large matrices.  JAX can invert linear operators, as discussed in that lecture.
